src/app/GitUI/Editor/ComboBoxWatermarkManager.cs

```diff
@@ -0,0 +1,185 @@
+        {
+            ShowWatermark();
+        }
+
+        _comboBox.Enter += ComboBox_Enter;
+        _comboBox.LostFocus += ComboBox_LostFocus;
+        _comboBox.TextChanged += ComboBox_TextChanged; // Handle programmatic changes or pasting
+        _comboBox.FontChanged += ComboBox_FontChanged; // Keep original font updated
+        _comboBox.ForeColorChanged += ComboBox_ForeColorChanged; // Keep original color updated
+    }
+
+    public string ComboBoxText
+    {
+        get
+        {
+            return _isWatermarkActive
+                ? string.Empty
+                : _comboBox.Text;
+        }
+    }
+
+    public bool WatermarkActive => _isWatermarkActive;
+
+    private void ComboBox_Enter(object sender, EventArgs e)
+    {
+        ActOpaqueToTextChanged(static wmm => wmm.HideWatermarkInternal(resetText: true));
+    }
+
+    private void ComboBox_LostFocus(object sender, EventArgs e)
+    {
+        if (string.IsNullOrEmpty(_comboBox.Text))
+        {
+            ActOpaqueToTextChanged(static wmm => wmm.ShowWatermark());
+        }
+    }
+
+    private void ComboBox_TextChanged(object sender, EventArgs e)
+    {
+        string comboBoxText = _comboBox.Text;
+
+        // If text changes while watermark is active (e.g., programmatically), hide watermark
+        if (_isWatermarkActive && comboBoxText != _watermarkText && comboBoxText != "")
+        {
+            ActOpaqueToTextChanged(static wmm => wmm.HideWatermarkInternal(resetText: false));
+        }
+        else if (!_isWatermarkActive && comboBoxText == "")
+        {
+            ActOpaqueToTextChanged(static wmm => wmm.ShowWatermark());
+        }
+    }
+
+    private void ComboBox_FontChanged(object sender, EventArgs e)
+    {
+        // If the font changes while the watermark is not active, update the original font.
+        if (!_isWatermarkActive && !_isDisposed && _comboBox.Font != _watermarkFont)
+        {
+            _originalFont = _comboBox.Font;
+
+            // Recreate watermark font based on new original font
+            _watermarkFont?.Dispose(); // Dispose previous watermark font if exists
+            _watermarkFont = new Font(_originalFont, FontStyle.Italic);
+        }
+    }
+
+    private void ComboBox_ForeColorChanged(object sender, EventArgs e)
+    {
+        // If the color changes while the watermark is not active, update the original color.
+        if (!_isWatermarkActive && !_isDisposed && _comboBox.ForeColor != _watermarkColor)
+        {
+            _originalForeColor = _comboBox.ForeColor;
+        }
+    }
+
+    /// <summary>
+    ///  Displays the watermark text in the ComboBox.
+    /// </summary>
+    public void ShowWatermark()
+    {
+        if (_isWatermarkActive || _isDisposed || _comboBox.Focused)
+        {
+            return;
+        }
+
+        _isWatermarkActive = true;
+        _comboBox.Font = _watermarkFont;
+        _comboBox.ForeColor = _watermarkColor;
+        _comboBox.Text = _watermarkText;
```

Comment:
How is this supposed to work? It notifies other subscribers of `TextChanged`. `ActOpaqueToTextChanged` does not address this issue.